### PR TITLE
Operators: elementwise multiplication with non-scalar tensors and typing

### PR DESCRIPTION
**Pull request by fzimmermann89**
We only supported (i.e. tested) for multiplication with a scalar tensor.

This adds tests and logic for elementwise multiplication with nd-tensors.

Related #247 

and adds missing type hints.

### CODE REVIEW

**Comment by fzimmermann89** [comment]
(ignore the merge, it will be squashed anyways...)

File: src/mrpro/operators/_LinearOperator.py [hunk 1]
```diff
@@ -82,8 +83,22 @@ def adjoint(self, x: torch.Tensor) -> tuple[torch.Tensor,]:
         return (self._operator1.adjoint(x)[0] + self._operator2.adjoint(x)[0],)
 
 
-class LinearOperatorElementwiseProduct(LinearOperator, OperatorElementwiseProduct[torch.Tensor, tuple[torch.Tensor,]]):
-    """Operator elementwise multiplication with scalar/tensor."""
+class LinearOperatorElementwiseProductRight(
+    LinearOperator, OperatorElementwiseProductRight[torch.Tensor, tuple[torch.Tensor,]]
+):
+    """Operator elementwise right multiplication with scalar/tensor."""
```
Comment (ckolbPTB) [inline src/mrpro/operators/_LinearOperator.py]: In the LinearOperator is the only time where you mention tensor and scalar as input. Nevertheless the signature only allows for tensors so I would suggest to remove scalar from the doc string.

File: tests/operators/test_operators.py [hunk 2]
```diff
@@ -128,80 +133,114 @@ def test_sum_operator_linearoperator():
...
+    b = torch.tensor(value)
     c = b * a
-    x = torch.arange(10)
+    x = RandomGenerator(0).complex64_tensor(10)
     (y1,) = c(x)
-    y2 = a(x)[0] * b
+    y2 = b * a(x)[0]
 
     torch.testing.assert_close(y1, y2)
     assert isinstance(c, Operator), 'Operator * scalar should be an Operator'
```
Comment (ckolbPTB) [inline tests/operators/test_operators.py]: To be super accurate this should be scaler * Operator and similarly for the other rproduct tests

**Review by ckolbPTB — APPROVED** [review]
Just a few small things, otherwise looks good.

**Comment by fzimmermann89** [comment]
Mhm, this is unfortunatly currently not a viable approach because of mypy and pytorch:

torch has the typehint
`torch.Tensor._mul__(self,other:Any)->Tensor`

even thought, __mul__ does not work with our operators.
so for Tensor*Operator at runtim, our __rmul__ gets  called.
But mypy thinks Tensor.__mul__ will be called and return a Tensor.
So static type of the return does not match the actual type.

I opened an issue at pytorch https://github.com/pytorch/pytorch/issues/124015 . But not sure when and if this will ever get addressed...

Our code is correct, as far as I know. So I would still add it. But we should keep this issue and not use Tensor*operator in mypy checked code ... ...  (Which is annoying, as this is the  case we want to use it for..)

**Pull request closed without merging**